Initial implementation of Argo workflows.

martinpovolny opened the pull request:
## Related Issues and Dependencies

https://github.com/aicoe-aiops/ocp-ci-analysis/issues/124

## This introduces a breaking change

- [ ] Yes
- [x] No

Changing the buckets, but those should be unused so far.

## This Pull Request implements

Initial implementation of Argo Workflows. Calling 3 notebooks to generate some metrics.

## Description

Details are in this issue: https://github.com/aicoe-aiops/ocp-ci-analysis/issues/124

## Testing
@oindrillac : Based on our discussion I am adding some notes on testing.

To test the generation of manifests:
```
cd manifests
kustomize build
```
Then review the output and/or pass it to `oc apply -f`

Running the notebooks take the command from the template. E.g.:
```
 ./notebook-executor.sh notebooks/data-sources/TestGrid/merics/persistent_failures_analysis.ipynb /tmp/persistent_failures_analysis.ipynb
```

The workflow can be tested by loading it into Argo manually using `argo` command line tool.

## FIXME:

~~Then `notebook-executor` does not set the current path to the location of the notebook. This means that the file path such as `"../../../../data/raw/testgrid_810.json.gz"` does not lead to the expected place.~~

sesheta commented:
<details>
<summary>Pre-Commit Test failed! Click here</summary>

```
[INFO] Initializing environment for git://github.com/Lucas-C/pre-commit-hooks.
[INFO] Initializing environment for git://github.com/pre-commit/pre-commit-hooks.
[INFO] Initializing environment for git://github.com/pycqa/pydocstyle.git.
[INFO] Initializing environment for https://github.com/pre-commit/pre-commit-hooks.
[INFO] Initializing environment for https://github.com/pre-commit/mirrors-mypy.
[INFO] Initializing environment for https://github.com/psf/black.
[INFO] Initializing environment for https://github.com/tomcatling/black-nb.
[INFO] Initializing environment for https://github.com/s-weigand/flake8-nb.
[INFO] Initializing environment for https://github.com/s-weigand/flake8-nb:pep8-naming.
[INFO] Installing environment for git://github.com/Lucas-C/pre-commit-hooks.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for git://github.com/pre-commit/pre-commit-hooks.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for git://github.com/pycqa/pydocstyle.git.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/pre-commit/pre-commit-hooks.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/pre-commit/mirrors-mypy.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/psf/black.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/tomcatling/black-nb.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/s-weigand/flake8-nb.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
Tabs remover.............................................................Passed
Trim Trailing Whitespace.................................................Passed
Check for merge conflicts................................................Passed
Fix End of Files.........................................................Passed
Tests should end in _test.py.........................(no files to check)Skipped
check BOM - deprecated: use fix-byte-order-marker........................Passed
Check for case conflicts.................................................Passed
Check docstring is first.................................................Passed
Check JSON...............................................................Passed
Check for broken symlinks............................(no files to check)Skipped
Detect Private Key.......................................................Passed
Check python ast.........................................................Passed
Debug Statements (Python)................................................Passed
pydocstyle...............................................................Passed
Check Toml...............................................................Passed
Check Yaml...............................................................Failed
- hook id: check-yaml
- exit code: 1

expected a single document in the stream
  in "manifests/buckets.yaml", line 1, column 1
but found another document
  in "manifests/buckets.yaml", line 11, column 1

Fix End of Files.........................................................Passed
Trim Trailing Whitespace.................................................Passed
mypy.....................................................................Passed
black....................................................................Passed
black-nb.................................................................Passed
flake8-nb................................................................Passed
```
</details>

sesheta commented:
<details>
<summary>Pre-Commit Test failed! Click here</summary>

```
[INFO] Initializing environment for git://github.com/Lucas-C/pre-commit-hooks.
[INFO] Initializing environment for git://github.com/pre-commit/pre-commit-hooks.
[INFO] Initializing environment for git://github.com/pycqa/pydocstyle.git.
[INFO] Initializing environment for https://github.com/pre-commit/pre-commit-hooks.
[INFO] Initializing environment for https://github.com/pre-commit/mirrors-mypy.
[INFO] Initializing environment for https://github.com/psf/black.
[INFO] Initializing environment for https://github.com/tomcatling/black-nb.
[INFO] Initializing environment for https://github.com/s-weigand/flake8-nb.
[INFO] Initializing environment for https://github.com/s-weigand/flake8-nb:pep8-naming.
[INFO] Installing environment for git://github.com/Lucas-C/pre-commit-hooks.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for git://github.com/pre-commit/pre-commit-hooks.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for git://github.com/pycqa/pydocstyle.git.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/pre-commit/pre-commit-hooks.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/pre-commit/mirrors-mypy.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/psf/black.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/tomcatling/black-nb.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/s-weigand/flake8-nb.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
Tabs remover.............................................................Passed
Trim Trailing Whitespace.................................................Passed
Check for merge conflicts................................................Passed
Fix End of Files.........................................................Passed
Tests should end in _test.py.........................(no files to check)Skipped
check BOM - deprecated: use fix-byte-order-marker........................Passed
Check for case conflicts.................................................Passed
Check docstring is first.................................................Passed
Check JSON...............................................................Passed
Check for broken symlinks............................(no files to check)Skipped
Detect Private Key.......................................................Passed
Check python ast.........................................................Passed
Debug Statements (Python)................................................Passed
pydocstyle...............................................................Passed
Check Toml...............................................................Passed
Check Yaml...............................................................Failed
- hook id: check-yaml
- exit code: 1

expected a single document in the stream
  in "manifests/buckets.yaml", line 1, column 1
but found another document
  in "manifests/buckets.yaml", line 11, column 1

Fix End of Files.........................................................Passed
Trim Trailing Whitespace.................................................Passed
mypy.....................................................................Passed
black....................................................................Passed
black-nb.................................................................Passed
flake8-nb................................................................Passed
```
</details>

martinpovolny commented:
/retest

tumido commented on manifests/buckets.yaml:
```diff
@@ -3,7 +3,7 @@ kind: ObjectBucketClaim
 metadata:
   name: black-flake
```
you'd need to temporarily change this name (or disable the ArgoCD autosync, delete the claim manually, wait for the operator to reconcile and then apply this as a new claim)

The bucket name can't be changed on the fly.

martinpovolny replied:
Yes, I know about that problem. Thx.

tumido commented on manifests/wftmpl.yaml:
```diff
@@ -1,63 +1,66 @@
-    #           parameters: [{name: notebook, value: <PATH_WITHIN_NOTEBOOKS_FOLDER>}]
-    #       - ...
+    - name: entrypoint
+      steps:
+        - - name: metrics
+            template: metrics
+
+    - name: metrics
+      dag:
+        tasks:
+          - name: number-of-flakes
+            template: notebook-executor
+            arguments:
+              parameters:
+                [{ name: notebook, value: data-sources/TestGrid/number_of_flakes.ipynb }]
+
+          - name: persistent-failures
+            template: notebook-executor
+            arguments:
+              parameters:
+                [{ name: notebook, value: data-sources/TestGrid/persistent_failures_analysis.ipynb}]
+
+          - name: test-pass-failures
+            template: notebook-executor
+            arguments:
+              parameters:
+                [{ name: notebook, value: data-sources/TestGrid/test_pass_failures.ipynb}]
```
You don't need a dag + steps if you're just executing things in parallel... You can do it via steps directly and in a readable way:

https://argoproj.github.io/argo-workflows/workflow-concepts/#steps

martinpovolny replied:
I expect to have dependencies between steps in the near future. My understanding is that a dag w/o dependencies will work as needed for now.

tumido replied:
yes. The dag without dependencies will execute all the tasks in parallel - if that's the intended behavior, than it's fine.

I would choose an approach of keeping it tidy (having just 3 parallel steps like this for now and expand on it in future rather than preparing for something that might or might not happen.

```yaml
      steps:
        - - name: number-of-flakes
            template: notebook-executor
            arguments:
              parameters:
                [{ name: notebook, value: data-sources/TestGrid/number_of_flakes.ipynb }]

          - name: persistent-failures
            template: notebook-executor
            arguments:
              parameters:
                [{ name: notebook, value: data-sources/TestGrid/persistent_failures_analysis.ipynb}]

          - name: test-pass-failures
            template: notebook-executor
            arguments:
              parameters:
                [{ name: notebook, value: data-sources/TestGrid/test_pass_failures.ipynb}]
```

But that's your call. It's now wrong either way. :slightly_smiling_face:

martinpovolny replied:
Ok, changed as you wrote.

martinpovolny commented:
/deploy

harshad16 commented:
@martinpovolny as the build-check has failed for the pull request 
the deploy command would also fail in building , we should first try to fix the source code

harshad16 commented:
/retest

sesheta commented:
Hi @harshad16,
 you are not authorised to run `/retest` command.
 Please contact a MEMBER.

harshad16 commented:
/retest

martinpovolny commented:
Yes, but I do not see the output from the "The image build test failed!" I get all kinds of different errors when I search for it.

/retest

harshad16 commented:
/retest

harshad16 commented:
The build-check is failing in this current PR, due to :
```
Pipfile.lock hash '\''9633df03d17d8a254e6d737ca6b45cb7a00cb7b3c792d20bf0741c4d6c9460c8'\'' does not correspond to hash computed based on Pipfile '\''7b7d14f1290aad2dd7314bb24b793ba4dd775265e43d118496c6f37f3e5d1cde'\'', aborting deployment
```
The pipfile was updated with a new package, but the pipfile.lock was not updated which causes the build-check to fail.

steps to view the issue:
- click on the build-check `details` link
- in the tekton pipeline, goto : `> pr-build-run > run-build`
- scroll to the bottom.

martinpovolny commented:
Thanks, @harshad16 

I suspected something like that therefore I put the `/retest` in the comment: https://github.com/aicoe-aiops/ocp-ci-analysis/pull/131#issuecomment-781392697 but did not see a build being attempted after that.

Now it's expired again so I am doing a

/retest

again and will check again after breakfast ;-)

martinpovolny commented:
This time the result is really there, no gateway timeouts or other errors and I get the report. But if I did not get your explanation of what was wrong I would be lost.

The red things seem to be irrelevant while the problem you pointed out is actually green. One needs to be an expert in parsing the output to figure what is wrong.
![figuring-what-went-wrong](https://user-images.githubusercontent.com/51095/108473153-f1588e80-728d-11eb-9766-1b0d82342ab7.jpg)


See the screenshot below

martinpovolny commented:
With help of @tumido 's issues:

https://github.com/aicoe-aiops/mailing-list-analysis-toolkit/issues/51
https://github.com/aicoe-aiops/mailing-list-analysis-toolkit/issues/46
https://github.com/thoth-station/thoth-application/issues/816

Btw great read for people who do not quite understand `pip`, `pipenv`, and `pipenv install`, `pipenv install .`  .... 
Much better than https://pipenv-fork.readthedocs.io/en/latest/basics.html 
and with
https://github.com/aicoe-aiops/mailing-list-analysis-toolkit/blob/master/Pipfile#L21-L25

I moved forward and added the Pipfile.lock changes. Thx!

martinpovolny commented:
@oindrillac : More on testing w/o MOC cluster.

I have been able to get Argo working here: http://argo-server-opf-argo.apps.mpovtest.lab.upshift.rdu2.redhat.com/workflows/opf-argo. It runs simple/sample workflows sucessfully.

But I do not have storage (buckets) working. And that we need for something serious.

I have an ArgoCD app to install Ceph here: https://argocd-server-aicoe-argocd-dev.apps.mpovtest.lab.upshift.rdu2.redhat.com/applications/ceph
but it's not synchronized and I do not see a meaningful error message in the UI. Maybe some expert can help with this testing workflow? @HumairAK?

@tumido gave me some hints about "local storage". But it was too much dark magic for me at this point.

harshad16 commented:
> This time the result is really there, no gateway timeouts or other errors and I get the report. But if I did not get your explanation of what was wrong I would be lost.
> I am sure a python expert with a knowledge of Tekton UI can figure that out. But probably not a data scientist or me.

Thanks for the feedback, will loop this in the aicoe-ci, try to get better documentation for this.
And will also check on how to better develop this.

martinpovolny commented:
Regarding the idea of testing this on Quicklab I opened this issue: https://github.com/operate-first/apps/issues/279 . Currently I am trying to get Ceph working.
